explorer/interpreter/type_checker.cpp

```diff
@@ -5291,57 +5291,54 @@ auto TypeChecker::DeclareClassDeclaration(Nonnull<ClassDeclaration*> class_decl,
+  // hasn't been typechecked yet and therefore doesn't have its `static_type`
+  // set.
+
+  // TODO: Verify just that is before any data member declarations, and there is
+  // at most one, and delay remaining work (type checking, base class
+  // evaluation, etc.) until the `extend base` declaration is processed in
+  // order.
   std::optional<Nonnull<const NominalClassType*>> base_class;
-  bool after_data_member = false;
+  bool first = true;
   for (Nonnull<Declaration*> m : class_decl->members()) {
-    switch (m->kind()) {
-      case DeclarationKind::VariableDeclaration:
-      case DeclarationKind::MixDeclaration: {
-        after_data_member = true;
-        break;
-      }
-      case DeclarationKind::ExtendBaseDeclaration: {
-        if (after_data_member) {
-          return ProgramError(m->source_loc())
-                 << "`extend base:` declaration must not be after `var` or "
-                    "`mix` declarations in a class.";
-        }
-        if (base_class.has_value()) {
-          return ProgramError(m->source_loc())
-                 << "At most one `extend base:` declaration in a class.";
-        }
-        Nonnull<Expression*> base_class_expr =
-            cast<ExtendBaseDeclaration>(*m).base_class();
-        CARBON_ASSIGN_OR_RETURN(const auto base_type,
-                                TypeCheckTypeExp(base_class_expr, class_scope));
-        if (base_type->kind() != Value::Kind::NominalClassType) {
-          return ProgramError(m->source_loc())
-                 << "Unsupported base class type for class `"
-                 << class_decl->name()
-                 << "`. Only simple classes are currently supported as base "
-                    "class.";
-        }
-        CARBON_RETURN_IF_ERROR(ExpectCompleteType(base_class_expr->source_loc(),
-                                                  "base class declaration",
-                                                  base_type));
-
-        base_class = cast<NominalClassType>(base_type);
-        if (base_class.value()->declaration().extensibility() ==
-            ClassExtensibility::None) {
-          return ProgramError(m->source_loc())
-                 << "Base class `" << base_class.value()->declaration().name()
-                 << "` is `final` and cannot be inherited. Add the `base` or "
-                    "`abstract` class prefix to `"
-                 << base_class.value()->declaration().name()
-                 << "` to allow it to be inherited";
-        }
-        class_decl->set_base_type(base_class);
-        break;
+    if (m->kind() != DeclarationKind::ExtendBaseDeclaration) {
+      first = false;
+    } else if (base_class.has_value()) {
+      return ProgramError(m->source_loc())
+             << "At most one `extend base:` declaration in a class.";
+    } else if (!first) {
+      return ProgramError(m->source_loc())
+             << "`extend base:` declaration must first in a class.";
```

Comment:
It'd be nice to leave a breadcrumb that this is a limitation of the Explorer, not a language rule.

```suggestion
             << "`extend base:` declarations after the first declaration in the class are not yet supported";
```

Reply:
Agreed